python/ray/data/dataset.py

```diff
@@ -2311,6 +2321,52 @@ def train_test_split(
                 )
             return ds.split_at_indices([ds_length - test_size])
 
+    def _stratified_split(
+        self, ds: "Dataset", test_size: Union[int, float], stratify: str
+    ) -> Tuple["MaterializedDataset", "MaterializedDataset"]:
+        """Perform stratified train-test split on the dataset.
+        
+        Args:
+            ds: The dataset to split.
+            test_size: Test size as int or float.
+            stratify: Column name to use for stratified sampling.
+            
+        Returns:
+            Train and test subsets as two MaterializedDatasets.
+        """
+        # Normalize test_size to float (only materialize if needed)
+        if isinstance(test_size, int):
+            ds_length = ds.count()
+            if test_size <= 0 or test_size >= ds_length:
+                raise ValueError(
+                    "If `test_size` is an int, it must be bigger than 0 and smaller "
+                    f"than the size of the dataset ({ds_length}). "
+                    f"Got {test_size}."
+                )
+            test_size = test_size / ds_length
+        
+        if test_size <= 0 or test_size >= 1:
+            raise ValueError(
+                "For stratified splitting, test_size must be between 0 and 1 "
+                f"(or equivalent int). Got {test_size}."
+            )
+        
+        def add_train_flag(group_batch):
+            import pyarrow as pa
+            # group_batch is already a pandas DataFrame
+            df = group_batch
+            n = len(df)
+            test_count = int(n * test_size)
+            df["is_train"] = [True] * (n - test_count) + [False] * test_count
+            return pa.Table.from_pandas(df)
```

Comment:
I think the default batch format should be numpy, not pandas
And we don't need to manually convert the batch to a pyarrow table. This UDF is supposed to return a batch.
you can just do this:
```
            n = len(group_batch)
            test_count = int(n * test_size)
            group_batch["is_train"] = [True] * (n - test_count) + [False] * test_count
             return group_batch
```